substrate/frame/stake-tracker/src/lib.rs

```diff
@@ -0,0 +1,347 @@
+	//
+	// Note: it is assumed that who's staking state is updated *before* the caller calling into
+	// this method. Thus, the nominations before the nominator has been removed from staking are
+	// passed in, so that the target list can be updated accordingly.
+	fn on_nominator_remove(who: &T::AccountId, nominations: Vec<T::AccountId>) {
+		let nominator_vote = Self::active_vote_of(who);
+
+		// updates the nominated target's score. Note: this may update the score of up to
+		// `T::MaxNominations` validators.
+		for t in nominations.iter() {
+			Self::update_score::<T::TargetList>(&t, StakeImbalance::Negative(nominator_vote))
+		}
+
+		let _ = T::VoterList::on_remove(&who).defensive_proof(
+			"the nominator exists in the list as per the contract with staking; qed.",
+		);
+	}
+
+	// Fired when someone removes their intention to validate, either due to chill or nominating.
+	fn on_validator_remove(who: &T::AccountId) {
+		let _ = T::TargetList::on_remove(&who).defensive_proof(
+			"the validator exists in the list as per the contract with staking; qed.",
+		);
+
+		// validator is also a nominator.
+		Self::on_nominator_remove(who, vec![]);
+	}
+
+	// Fired when an existing nominator updates their nominations.
+	//
+	// This is called when a nominator updates their nominations. The nominator's stake remains the
+	// same (updates to the nominator's stake should emit [`Self::on_stake_update`] instead).
+	// However, the score of the nominated targets must be updated accordingly.
+	//
+	// Note: it is assumed that who's staking state is updated *before* calling this method.
+	fn on_nominator_update(who: &T::AccountId, prev_nominations: Vec<T::AccountId>) {
+		let nominator_vote = Self::active_vote_of(who);
+		let curr_nominations =
+			<T::Staking as StakingInterface>::nominations(&who).unwrap_or_default();
+
+		// new nominations
+		for target in curr_nominations.iter() {
+			if !prev_nominations.contains(target) {
+				Self::update_score::<T::TargetList>(
+					&target,
+					StakeImbalance::Positive(nominator_vote),
+				);
+			}
+		}
+		// removed nominations
+		for target in prev_nominations.iter() {
+			if !curr_nominations.contains(target) {
+				Self::update_score::<T::TargetList>(
+					&target,
+					StakeImbalance::Negative(nominator_vote),
+				);
+			}
+		}
+	}
+
+	// noop: the updates to target and voter lists when applying a slash are performed
+	// through [`Self::on_nominator_remove`] and [`Self::on_validator_remove`] when the stakers are
+	// chilled. When the slash is applied, the ledger is updated, thus the stake is propagated
+	// through the `[Self::update::<T::EventListener>]`.
+	fn on_slash(
+		_stash: &T::AccountId,
+		_slashed_active: BalanceOf<T>,
+		_slashed_unlocking: &BTreeMap<sp_staking::EraIndex, BalanceOf<T>>,
+		_slashed_total: BalanceOf<T>,
+	) {
+		#[cfg(any(std, no_std))]
+		frame_support::defensive!("unexpected call to OnStakingUpdate::on_slash");
```

Comment:
Also, this seems like a wrong assumption here. Staking will emit slash events here as well https://github.com/paritytech/substrate/pull/14527

Reply:
Yes, that's correct. I removed the defensive call from here. However, the `OnStakeUpdate::on_slash` will remain a noop, as the `ledger.update()` call on validators and nominators' ledgers after the slashes are applied will take care of updating the target list scores as expected.